noir-projects/aztec-nr/aztec/src/state_vars/private_set.nr

```diff
@@ -67,40 +55,42 @@ where
         self,
         options: NoteGetterOptions<Note, N, PREPROCESSOR_ARGS, FILTER_ARGS>,
     ) -> BoundedVec<Note, MAX_NOTE_HASH_READ_REQUESTS_PER_CALL> {
-        let (notes, note_hashes) = get_notes(self.context, self.storage_slot, options);
+        let (retrieved_notes, note_hashes) = get_notes(self.context, self.storage_slot, options);
         // We iterate in a range 0..options.limit instead of 0..notes.len() because options.limit is known at compile
         // time and hence will result in less constraints when set to a lower value than
         // MAX_NOTE_HASH_READ_REQUESTS_PER_CALL.
         for i in 0..options.limit {
-            if i < notes.len() {
-                let note = notes.get_unchecked(i);
+            if i < retrieved_notes.len() {
+                let retrieved_note = retrieved_notes.get_unchecked(i);
                 let note_hash = note_hashes.get_unchecked(i);
                 // We immediately destroy the note without doing any of the read request checks `remove` typically
                 // performs because we know that the `get_notes` call has already placed those constraints.
-                destroy_note_unsafe(self.context, note, note_hash);
+                destroy_note_unsafe(self.context, retrieved_note, note_hash);
             }
         }
 
-        notes
+        // Since the notes were already destroyed, we no longer need the additional data in RetrievedNote and can
+        // discard it.
+        retrieved_notes.map(|retrieved_note| retrieved_note.note)
     }
 
     /// Note that if you obtained the note via `get_notes` it's much better to use `pop_notes` as `pop_notes` results
     /// in significantly less constrains due to avoiding an extra hash and read request check.
-    pub fn remove(self, note: Note) {
-        let note_hash = compute_note_hash_for_read_request(note, self.storage_slot);
+    pub fn remove(self, retrieved_note: RetrievedNote<Note>) {
+        let note_hash = compute_note_hash_for_read_request(retrieved_note, self.storage_slot);
         let has_been_read =
             self.context.note_hash_read_requests.any(|r: ReadRequest| r.value == note_hash);
         assert(has_been_read, "Can only remove a note that has been read from the set.");
 
-        destroy_note_unsafe(self.context, note, note_hash);
+        destroy_note_unsafe(self.context, retrieved_note, note_hash);
     }
 
     /// Note that if you later on remove the note it's much better to use `pop_notes` as `pop_notes` results
     /// in significantly less constrains due to avoiding 1 read request check.
     pub fn get_notes<PREPROCESSOR_ARGS, FILTER_ARGS>(
         self,
         options: NoteGetterOptions<Note, N, PREPROCESSOR_ARGS, FILTER_ARGS>,
-    ) -> BoundedVec<Note, MAX_NOTE_HASH_READ_REQUESTS_PER_CALL> {
+    ) -> BoundedVec<RetrievedNote<Note>, MAX_NOTE_HASH_READ_REQUESTS_PER_CALL> {
```

Comment:
Interesting. This already means that you can only really call `remove` if you got the note from `get_notes`, which is an interesting property.